File: CPP/No137.cpp
/**
 * Created by Xiaozhong on 2020/8/13.
 * Copyright (c) 2020/8/13 Xiaozhong. All rights reserved.
 */
#include "vector"
#include "set"

using namespace std;

class Solution {
public:
    int singleNumber(vector<int> &nums) {
        set<long> num_set;
        long sum_set = 0, sum_array = 0;
        for (int i : nums) {
            sum_array += i;
            num_set.insert(i);
        }
        for (long l : num_set) sum_set += l;
        // 3 * (a + b + c) - 3 * a - 3 * b = 2 * c
        return (int) ((3 * sum_set - sum_array) / 2);
    }
};

File: Java/src/No137.java
/*
 * Copyright (c) 2021, HuXiaozhong. All rights reserved.
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
 */

import java.util.HashSet;
import java.util.Set;

public class No137 {
    public int singleNumber(int[] nums) {
        Set<Long> set = new HashSet<>();
        long sum = 0;
        for (long num : nums) {
            sum += num;
            set.add(num);
        }
        long sumSet = 0;
        for (long num : set) sumSet += num;
        return (int) ((3 * sumSet - sum) / 2);
    }

    public static void main(String[] args) {
        int[] nums = {43, 16, 45, 89, 45, -2147483648, 45, 2147483646, -2147483647, -2147483648, 43, 2147483647, -2147483646, -2147483648, 89, -2147483646, 89, -2147483646, -2147483647, 2147483646, -2147483647, 16, 16, 2147483646, 43};
        No137 no137 = new No137();
        System.out.println(no137.singleNumber(nums));
    }
}
